kratos/utilities/binbased_fast_point_locator.h

```diff
@@ -203,132 +208,80 @@ class BinBasedFastPointLocator
      * @param pElement The pointer to the element containing the located point
+     * @param ItResultBegin The iterator of the search
      * @param MaxNumberOfResults The max number of results to be considered
      * @param Tolerance The tolerance considered on the search
      * @return If "false" is devolved the element is not found
      * @note this function is threadsafe and can be used within OpenMP loops
      */
-    bool FindPointOnMeshSimplified(
+    bool FindPointOnMesh(
         const array_1d<double, 3 >& rCoordinates,
         Vector& rNShapeFunction,
         Element::Pointer& pElement,
+        ResultIteratorType ItResultBegin,
         const SizeType MaxNumberOfResults = 1000,
         const double Tolerance = 1.0e-5
         )
     {
-        ResultContainerType results(MaxNumberOfResults);
-
-        if(rNShapeFunction.size() != TDim+1) {
-            rNShapeFunction.resize(TDim+1,false);
-        }
-
-        array_1d<double,TDim+1> aux;
-
-        const bool is_found = FindPointOnMesh(rCoordinates, aux, pElement, results.begin(), MaxNumberOfResults, Tolerance);
-
-        if(is_found) {
-            noalias(rNShapeFunction) = aux;
-        }
+        // Ask to the container for the list of candidate elements
+        SizeType results_found = mpBinsObjectDynamic->SearchObjectsInCell(rCoordinates, ItResultBegin, MaxNumberOfResults);
 
-        return is_found;
-    }
-    
-    //***************************************
-    //***************************************
-
-    inline bool CalculatePosition(
-        GeometryType& geom,
-        const double xc, 
-        const double yc, 
-        const double zc,
-        array_1d<double, 3 > & N,
-        const double Tolerance = 1.0e-5
-        )
-    {
-        const double x0 = geom[0].X();
-        const double y0 = geom[0].Y();
-        const double x1 = geom[1].X();
-        const double y1 = geom[1].Y();
-        const double x2 = geom[2].X();
-        const double y2 = geom[2].Y();
-
-        const double area = CalculateVol(x0, y0, x1, y1, x2, y2);
-
-        double inv_area = 0.0;
-        if (area == 0.0) {
-            KRATOS_ERROR << "Element with zero area found with the current geometry " << geom << std::endl;
-        } else {
-            inv_area = 1.0 / area;
-        }
+        if (results_found > 0) {
+            // Loop over the candidate elements and check if the particle falls within
+            for (IndexType i = 0; i < results_found; i++) {
+                GeometryType& geom = (*(ItResultBegin + i))->GetGeometry();
 
-        N[0] = CalculateVol(x1, y1, x2, y2, xc, yc) * inv_area;
-        N[1] = CalculateVol(x2, y2, x0, y0, xc, yc) * inv_area;
-        N[2] = CalculateVol(x0, y0, x1, y1, xc, yc) * inv_area;
+                // Find local position
+                array_1d<double, 3> point_local_coordinates;
+                const bool is_found = geom.IsInside(rCoordinates, point_local_coordinates, Tolerance);
```

Comment:
Thanks @loumalouomega So you are generalizing the point locator by using the IsInside function right... I will try to see whether I can implement the PointLocalCoordinates for Quad2D. But thanks for the reference, will take a look at it.